File: sdks/java/src/main/java/org/byteveda/taskito/webhooks/DeliveryStore.java
package org.byteveda.taskito.webhooks;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.type.CollectionType;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import org.byteveda.taskito.Taskito;
import org.byteveda.taskito.errors.WebhookException;

/**
 * Per-subscription webhook delivery log, persisted in the queue's settings KV.
 *
 * <p>Each subscription gets its own JSON list under
 * {@code webhooks:deliveries:<subscriptionId>}, append-only with FIFO eviction
 * once the per-webhook cap is hit — enough to debug recent activity without
 * unbounded growth. Records are stored oldest-first; {@link #listFor} reverses
 * for newest-first paging.
 */
final class DeliveryStore {
    static final String KEY_PREFIX = "webhooks:deliveries:";
    private static final int MAX_PER_WEBHOOK = 200;
    private static final ObjectMapper JSON = new ObjectMapper();

    private final Taskito queue;

    DeliveryStore(Taskito queue) {
        this.queue = queue;
    }

    /** Append {@code delivery} to its subscription's log, trimming to the cap. */
    void record(Delivery delivery) {
        List<Delivery> rows = load(delivery.subscriptionId());
        rows.add(delivery);
        if (rows.size() > MAX_PER_WEBHOOK) {
            rows = new ArrayList<>(rows.subList(rows.size() - MAX_PER_WEBHOOK, rows.size()));
        }
        save(delivery.subscriptionId(), rows);
    }

    /** Newest-first, optionally filtered by status/event, then paged. */
    List<Delivery> listFor(String subscriptionId, String statusFilter, String eventFilter, int limit, int offset) {
        List<Delivery> rows = load(subscriptionId);
        List<Delivery> out = new ArrayList<>();
        for (int i = rows.size() - 1; i >= 0; i--) {
            Delivery row = rows.get(i);
            if (statusFilter != null && !statusFilter.equals(row.status())) {
                continue;
            }
            if (eventFilter != null && !eventFilter.equals(row.event())) {
                continue;
            }
            out.add(row);
        }
        int from = Math.min(Math.max(offset, 0), out.size());
        int to = limit < 0 ? out.size() : Math.min(from + limit, out.size());
        return new ArrayList<>(out.subList(from, to));
    }

    Optional<Delivery> get(String subscriptionId, String deliveryId) {
        return load(subscriptionId).stream()
                .filter(row -> row.id().equals(deliveryId))
                .findFirst();
    }

    /** Drop the whole log for a subscription (called when the webhook is deleted). */
    void deleteFor(String subscriptionId) {
        queue.deleteSetting(KEY_PREFIX + subscriptionId);
    }

    private List<Delivery> load(String subscriptionId) {
        return queue.getSetting(KEY_PREFIX + subscriptionId)
                .map(DeliveryStore::parse)
                .orElseGet(ArrayList::new);
    }

    private void save(String subscriptionId, List<Delivery> rows) {
        try {
            queue.setSetting(KEY_PREFIX + subscriptionId, JSON.writeValueAsString(rows));
        } catch (Exception e) {
            throw new WebhookException("failed to persist webhook deliveries", e);
        }
    }

    private static List<Delivery> parse(String json) {
        try {
            CollectionType type = JSON.getTypeFactory().constructCollectionType(List.class, Delivery.class);
            return JSON.readValue(json, type);
        } catch (Exception e) {
            // A corrupt log must not wedge the dashboard — start fresh.
            return new ArrayList<>();
        }
    }
}

File: sdks/python/taskito/dashboard/delivery_store.py
"""Persistent webhook delivery log.

Each subscription gets its own JSON list under the key
``webhooks:deliveries:{subscription_id}`` in the ``dashboard_settings``
table. The store is append-only with FIFO eviction once the per-webhook
cap is hit (default 200 entries) — enough to debug recent activity
without unbounded growth.

The structure:

    [
      {
        "id": "uuid",
        "subscription_id": "sub-uuid",
        "event": "job.completed",
        "task_name": "send_email" | null,
        "job_id": "abc123" | null,
        "payload": {...},
        "status": "pending" | "delivered" | "failed" | "dead",
        "attempts": 3,
        "response_code": 200 | null,
        "response_body": "..." | null,
        "latency_ms": 42,
        "error": "..." | null,
        "created_at": 1234567890000,
        "completed_at": 1234567890420
      },
      ...
    ]

Records are inserted in chronological order; listing reverses for newest-first.
"""

from __future__ import annotations

import enum
import json
import logging
import time
import uuid
from dataclasses import asdict, dataclass, field
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from taskito.app import Queue


class DeliveryStatus(str, enum.Enum):
    """Terminal (or pending) state of a webhook delivery attempt.

    ``PENDING`` is a Python-only starting state — a record exists before its
    first attempt settles. Other SDKs record only settled attempts.
    """

    PENDING = "pending"
    DELIVERED = "delivered"
    FAILED = "failed"
    DEAD = "dead"


def _parse_status(value: Any) -> DeliveryStatus:
    """Read a stored status, falling back to ``PENDING`` on an unknown value."""
    try:
        return DeliveryStatus(str(value))
    except ValueError:
        return DeliveryStatus.PENDING


DELIVERY_PREFIX = "webhooks:deliveries:"
DEFAULT_MAX_PER_WEBHOOK = 200
RESPONSE_BODY_MAX_BYTES = 2048

logger = logging.getLogger("taskito.dashboard.deliveries")


@dataclass
class DeliveryRecord:
    """A single attempted webhook delivery."""

    id: str
    subscription_id: str
    event: str
    payload: dict[str, Any]
    task_name: str | None = None
    job_id: str | None = None
    status: DeliveryStatus = DeliveryStatus.PENDING
    attempts: int = 0
    response_code: int | None = None
    response_body: str | None = None
    latency_ms: int | None = None
    error: str | None = None
    created_at: int = field(default_factory=lambda: int(time.time() * 1000))
    completed_at: int | None = None

    @classmethod
    def from_row(cls, row: dict[str, Any]) -> DeliveryRecord:
        return cls(
            id=str(row["id"]),
            subscription_id=str(row["subscription_id"]),
            event=str(row["event"]),
            payload=dict(row.get("payload") or {}),
            task_name=row.get("task_name"),
            job_id=row.get("job_id"),
            status=_parse_status(row.get("status")),
            attempts=int(row.get("attempts", 0)),
            response_code=row.get("response_code"),
            response_body=row.get("response_body"),
            latency_ms=row.get("latency_ms"),
            error=row.get("error"),
            created_at=int(row.get("created_at", 0)),
            completed_at=row.get("completed_at"),
        )


def _new_id() -> str:
    return uuid.uuid4().hex


def _now_ms() -> int:
    return int(time.time() * 1000)


def _truncate(body: str | None, *, max_bytes: int = RESPONSE_BODY_MAX_BYTES) -> str | None:
    if body is None:
        return None
    encoded = body.encode("utf-8", errors="replace")
    if len(encoded) <= max_bytes:
        return body
    return encoded[:max_bytes].decode("utf-8", errors="replace") + "…"


class DeliveryStore:
    """List/insert/update delivery records keyed by subscription id."""

    def __init__(self, queue: Queue, *, max_per_webhook: int = DEFAULT_MAX_PER_WEBHOOK) -> None:
        self._queue = queue
        self._max = max_per_webhook

    # ── Internal ────────────────────────────────────────────────

    def _key(self, subscription_id: str) -> str:
        return DELIVERY_PREFIX + subscription_id

    def _load(self, subscription_id: str) -> list[dict[str, Any]]:
        raw = self._queue.get_setting(self._key(subscription_id))
        if not raw:
            return []
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            logger.warning("delivery log for %s is corrupt; resetting", subscription_id)
            return []
        return data if isinstance(data, list) else []

    def _save(self, subscription_id: str, rows: list[dict[str, Any]]) -> None:
        self._queue.set_setting(
            self._key(subscription_id),
            json.dumps(rows, separators=(",", ":")),
        )

    # ── Public API ─────────────────────────────────────────────

    def record_attempt(
        self,
        subscription_id: str,
        event: str,
        payload: dict[str, Any],
        *,
        status: DeliveryStatus,
        attempts: int,
        response_code: int | None = None,
        response_body: str | None = None,
        latency_ms: int | None = None,
        error: str | None = None,
        task_name: str | None = None,
        job_id: str | None = None,
    ) -> DeliveryRecord:
        """Append a delivery row and trim to the per-webhook cap."""
        now = _now_ms()
        record = DeliveryRecord(
            id=_new_id(),
            subscription_id=subscription_id,
            event=event,
            payload=payload,
            task_name=task_name,
            job_id=job_id,
            status=status,
            attempts=attempts,
            response_code=response_code,
            response_body=_truncate(response_body),
            latency_ms=latency_ms,
            error=error,
            created_at=now,
            completed_at=now if status is not DeliveryStatus.PENDING else None,
        )
        rows = self._load(subscription_id)
        rows.append(asdict(record))
        if len(rows) > self._max:
            rows = rows[-self._max :]
        self._save(subscription_id, rows)
        return record

    def list_for(
        self,
        subscription_id: str,
        *,
        status: DeliveryStatus | str | None = None,
        event: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> list[DeliveryRecord]:
        rows = list(reversed(self._load(subscription_id)))  # newest first
        if status:
            rows = [r for r in rows if r.get("status") == status]
        if event:
            rows = [r for r in rows if r.get("event") == event]
        page = rows[offset : offset + limit]
        return [DeliveryRecord.from_row(r) for r in page]

    def get(self, subscription_id: str, delivery_id: str) -> DeliveryRecord | None:
        for row in self._load(subscription_id):
            if row.get("id") == delivery_id:
                return DeliveryRecord.from_row(row)
        return None

    def delete_for(self, subscription_id: str) -> bool:
        return self._queue.delete_setting(self._key(subscription_id))

    def count_for(self, subscription_id: str) -> int:
        return len(self._load(subscription_id))
